text/0000-transition-hooks-events.md

```diff
@@ -0,0 +1,122 @@
+```
+
+The `willChangeTransition` hook should provide the same API surface as the `willTransition` hook, and the callback for the `willChangeTransition` event should be passed the same parameters as the callback for the `willTransition` event. This grants users wishing to subscribe to all transition events the ability to reuse one function across both hooks or both callbacks.
+
+```javascript
+function hookForAllTransitions(oldInfos, newInfos, transition) {
+  const ret = this.super(...arguments);
+
+  // do stuff for all transitions...
+
+  return ret;
+}
+
+function callbackForAllTransitions(transition) {
+  // do stuff for all transitions...
+}
+
+const Router = Ember.Router.extend({
+  willTransition: hookForAllTransitions,
+  willChangeTransition: hookForAllTransitions,
+});
+
+Router.on('willTransition', callbackForAllTransitions);
+Router.on('willChangeTransition', callbackForAllTransitions);
+```
+
+This will require changes to both Ember and tildeio/router.js.
+
+# How We Teach This
+
+> What names and terminology work best for these concepts and why?
+
+We propose an event name of `willChangeTransition` as parallel to the existing `willTransition` and `didTransition` hooks. It is understandable in context, and out of context while hewing close to existing patterns.
+
+> How is this idea best presented? As a continuation of existing Ember patterns, or as a wholly new one?
+
+This is new functionality but it was merely omitted from event handling in the existing implementation. There are already logical places inside of Ember to insert it, and the existing Ember patterns apply.
+
+> Would the acceptance of this proposal mean the Ember guides must be re-organized or altered? Does it change how Ember is taught to new users at any level?
+
+The guides would not need to be altered as this is a concept better suited for API-level documentation. New users will be unlikely to encounter this, and addon developers are likely the ones who most desire this feature.
+
+> How should this feature be introduced and taught to existing Ember users?
+
+The people who need this hook will seek it out. The target audience for this is implementors of addons as it generally will be used for changes across the entire application and not in one individual scenario. As such simply mentioning it in the Ember.js release blog entry and documenting it in the API will be sufficient.
+
+# Drawbacks
+
+> Why should we *not* do this? Please consider the impact on teaching Ember, on the integration of this feature with other existing and planned features, on the impact of the API churn on existing apps, etc.
+
+As this is a pure addition there will be no impacts to existing applications. There is a small increase in the public API of Ember, and a few more lines of code which we have to ship to application consumers. This is a relatively mature area of Ember that hasn't undergone significant change and as such it's unlikely to conflict with future plans.
```

Comment:
I don't expect any, but is there a possible performance impact when generating the arguments to each `willChangeTransition` call?